In [836]:
import pandas as pd
import re
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import warnings

pd.options.display.max_columns = None

warnings.filterwarnings('ignore')

ABRIMOS EL TXT 

In [837]:
data_txt = pd.read_csv('../datos/data_txt_clean.txt', sep= ';')
data_txt.head(2)
    

index         Q3                 Q4                       Q5          Q6  \
0      0      India  Bachelor’s degree                    Other  5-10 years   
1      1  Indonesia    Master’s degree  Program/Project Manager   20+ years   

       Q8                                                Q11        Q13  \
0  Python                                           A laptop  2-5 times   
1  Python  A cloud computing platform (AWS, Azure, GCP, h...      Never   

          Q15                        Q20                   Q21  Q22  \
0  5-10 years  Manufacturing/Fabrication      50-249 employees  3-4   
1    < 1 year  Manufacturing/Fabrication  1000-9,999 employees  1-2   

                                                 Q23            Q25  \
0                      No (we do not use ML methods)  25,000-29,999   
1  We are exploring ML methods (and may one day p...  60,000-69,999   

         Q26  Q33  Q35                                                Q41  \
0  $100-$999  NaN  NaN  Local development environments (RStudio, Jupyt...   
1  $0 ($USD)  NaN  NaN    Advanced statistical software (SPSS, SAS, etc.)   

                   Q7                           Q9                 Q12  \
0           Python, R                  Vim / Emacs  Google Cloud TPUs    
1   SQL, C, C++, Java   Notepad++,Jupyter Notebook                None   

                                                 Q14  \
0   Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ...   
1                                        Matplotlib    

                                 Q16  \
0  Scikit-learn , TensorFlow ,Caret    
1                      Scikit-learn    

                                                 Q17  \
0  Linear or Logistic Regression, Decision Trees ...   
1  Linear or Logistic Regression, Decision Trees ...   

                                                 Q24           Q32    Q34  
0   None of these activities are an important par...   PostgreSQL    None  
1   Build and/or run the data infrastructure that...           NaN    NaN

In [838]:
#Borramos la columna 'index' puesto que no es necesaria
data_txt.drop('index', inplace = True, axis = 1)
data_txt.head(2)

Q3                 Q4                       Q5          Q6      Q8  \
0      India  Bachelor’s degree                    Other  5-10 years  Python   
1  Indonesia    Master’s degree  Program/Project Manager   20+ years  Python   

                                                 Q11        Q13         Q15  \
0                                           A laptop  2-5 times  5-10 years   
1  A cloud computing platform (AWS, Azure, GCP, h...      Never    < 1 year   

                         Q20                   Q21  Q22  \
0  Manufacturing/Fabrication      50-249 employees  3-4   
1  Manufacturing/Fabrication  1000-9,999 employees  1-2   

                                                 Q23            Q25  \
0                      No (we do not use ML methods)  25,000-29,999   
1  We are exploring ML methods (and may one day p...  60,000-69,999   

         Q26  Q33  Q35                                                Q41  \
0  $100-$999  NaN  NaN  Local development environments (RStudio, Jupyt...   
1  $0 ($USD)  NaN  NaN    Advanced statistical software (SPSS, SAS, etc.)   

                   Q7                           Q9                 Q12  \
0           Python, R                  Vim / Emacs  Google Cloud TPUs    
1   SQL, C, C++, Java   Notepad++,Jupyter Notebook                None   

                                                 Q14  \
0   Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ...   
1                                        Matplotlib    

                                 Q16  \
0  Scikit-learn , TensorFlow ,Caret    
1                      Scikit-learn    

                                                 Q17  \
0  Linear or Logistic Regression, Decision Trees ...   
1  Linear or Logistic Regression, Decision Trees ...   

                                                 Q24           Q32    Q34  
0   None of these activities are an important par...   PostgreSQL    None  
1   Build and/or run the data infrastructure that...           NaN    NaN

In [839]:
data_txt.shape

(25973, 26)

ABRIMOS EL XML

In [840]:
data_xml = pd.read_xml('../datos/data_xml_clean.xml')
data_xml.sample(5)

level_0  index  time    age gender
176        176    176   412  30-34    Man
24375    24375  24375   752  22-24  Woman
811        811    811   519  40-44  Woman
17583    17583  17583  1976  50-54    Man
9932      9932   9932   558  50-54    Man

In [841]:
data_xml.drop('level_0', inplace = True, axis = 1)
data_xml.drop('index', inplace = True, axis = 1)
data_xml.head(2)

time    age gender
0   910  50-54    Man
1   784  50-54    Man

In [842]:
data_xml.shape

(25973, 3)

Como observamos que los 2 dataframes que son objeto de nuestro análisis tienen la misma longitud y cada longitud pertenece a un trabajador que ha respondido, decidimos unir todas las preguntas que han sido formuladas a cada usuario en un único dataframe y así trabajar con él.

In [843]:
df_unido = pd.concat([data_txt, data_xml], axis = 1 )
df_unido.head(2)

Q3                 Q4                       Q5          Q6      Q8  \
0      India  Bachelor’s degree                    Other  5-10 years  Python   
1  Indonesia    Master’s degree  Program/Project Manager   20+ years  Python   

                                                 Q11        Q13         Q15  \
0                                           A laptop  2-5 times  5-10 years   
1  A cloud computing platform (AWS, Azure, GCP, h...      Never    < 1 year   

                         Q20                   Q21  Q22  \
0  Manufacturing/Fabrication      50-249 employees  3-4   
1  Manufacturing/Fabrication  1000-9,999 employees  1-2   

                                                 Q23            Q25  \
0                      No (we do not use ML methods)  25,000-29,999   
1  We are exploring ML methods (and may one day p...  60,000-69,999   

         Q26  Q33  Q35                                                Q41  \
0  $100-$999  NaN  NaN  Local development environments (RStudio, Jupyt...   
1  $0 ($USD)  NaN  NaN    Advanced statistical software (SPSS, SAS, etc.)   

                   Q7                           Q9                 Q12  \
0           Python, R                  Vim / Emacs  Google Cloud TPUs    
1   SQL, C, C++, Java   Notepad++,Jupyter Notebook                None   

                                                 Q14  \
0   Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ...   
1                                        Matplotlib    

                                 Q16  \
0  Scikit-learn , TensorFlow ,Caret    
1                      Scikit-learn    

                                                 Q17  \
0  Linear or Logistic Regression, Decision Trees ...   
1  Linear or Logistic Regression, Decision Trees ...   

                                                 Q24           Q32    Q34  \
0   None of these activities are an important par...   PostgreSQL    None   
1   Build and/or run the data infrastructure that...           NaN    NaN   

   time    age gender  
0   910  50-54    Man  
1   784  50-54    Man

# EXPLORACION

Nuestro dataframe tiene más columnas de las que necesitamos para nuestro análisis, por tanto, eliminaremos las que no
necesitamos para tener un dataframe con las columnas que analizaremos

In [844]:
def eliminar_columnas(df):
    """
    Esta función elimina las columnas que no queremos tener en nuestro dataframe, puesto que no basaremos nuestro análisis en ellas
    Args: df
    df(dataframe):dataframe al que queremos aplicar la función
    """

    columnas_eliminadas = []
    patron = 'Q4|Q6|Q8|Q11|Q13|Q15|Q20|Q21|Q22|Q26|Q33|Q12|Q16|Q17|Q32|Q34|time|Q23|Q35'

    for col in df.columns:
        if re.findall(patron,col):
            columnas_eliminadas.append(col)
            
    eliminado = df.drop(columnas_eliminadas, axis=1, inplace = True)
    return eliminado

In [845]:
eliminar_columnas(df_unido)

In [846]:
col41 = data_txt[['Q41']]
col41

Q41
0      Local development environments (RStudio, Jupyt...
1        Advanced statistical software (SPSS, SAS, etc.)
2      Basic statistical software (Microsoft Excel, G...
3      Local development environments (RStudio, Jupyt...
4      Local development environments (RStudio, Jupyt...
...                                                  ...
25968  Business intelligence software (Salesforce, Ta...
25969                                                NaN
25970  Basic statistical software (Microsoft Excel, G...
25971                                                NaN
25972    Advanced statistical software (SPSS, SAS, etc.)

[25973 rows x 1 columns]

In [847]:
df = pd.concat([df_unido, col41], axis = 1 )
df.head(2)

Q3                       Q5            Q25                  Q7  \
0      India                    Other  25,000-29,999           Python, R   
1  Indonesia  Program/Project Manager  60,000-69,999   SQL, C, C++, Java   

                            Q9  \
0                  Vim / Emacs   
1   Notepad++,Jupyter Notebook   

                                                 Q14  \
0   Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ...   
1                                        Matplotlib    

                                                 Q24    age gender  \
0   None of these activities are an important par...  50-54    Man   
1   Build and/or run the data infrastructure that...  50-54    Man   

                                                 Q41  
0  Local development environments (RStudio, Jupyt...  
1    Advanced statistical software (SPSS, SAS, etc.)

In [848]:
#Cambiamos el nombre de nuestras columnas
df.rename(columns = {'Q3': 'country', 'Q5': 'current_role', 'Q25':'yearly_compensation', 'Q7': 'programming_languages', 'Q9' : 'ides', 'Q14': 'visualization', 'Q24' : 'activities', 'age' : 'age', 'gender' : 'gender', 'Q41' : 'primary_tool'}, inplace = True)

Comenzamos la exploración del nuevo dataframe, con el que trabajaremos a partir de ahora

In [849]:
df.head(3)

country             current_role yearly_compensation  \
0      India                    Other       25,000-29,999   
1  Indonesia  Program/Project Manager       60,000-69,999   
2   Pakistan        Software Engineer              $0-999   

  programming_languages                              ides  \
0             Python, R                       Vim / Emacs   
1     SQL, C, C++, Java        Notepad++,Jupyter Notebook   
2     Python, C++, Java  PyCharm ,Jupyter Notebook, Other   

                                       visualization  \
0   Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ...   
1                                        Matplotlib    
2                                        Matplotlib    

                                          activities    age gender  \
0   None of these activities are an important par...  50-54    Man   
1   Build and/or run the data infrastructure that...  50-54    Man   
2   None of these activities are an important par...  22-24    Man   

                                        primary_tool  
0  Local development environments (RStudio, Jupyt...  
1    Advanced statistical software (SPSS, SAS, etc.)  
2  Basic statistical software (Microsoft Excel, G...

In [850]:
#Creamos una función de limpieza básica para pasar a los diferentes dfs
def propiedades(df):
    """
    Esta función nos muestra la forma y el tamaño del dataframe y el tipo de dato para cada columna del dataframe
    Args: df
    df(dataframe):dataframe al que queremos aplicar la función
    """
    print("La forma es", df.shape)
    print('----------------------')
    print("El tamaño es", df.size)
    print('----------------------')
    print("El tipo de dato es", df.dtypes)

In [851]:
propiedades(df)

La forma es (25973, 10)
----------------------
El tamaño es 259730
----------------------
El tipo de dato es country                  object
current_role             object
yearly_compensation      object
programming_languages    object
ides                     object
visualization            object
activities               object
age                      object
gender                   object
primary_tool             object
dtype: object


Observamos que todas las columnas son de tipo object, y tenemos duda sobre si tenemos que cambiar la columna 'yearly_compensation' a una columna numérica o la columna 'age'

In [852]:
df['country'].unique()

array(['India', 'Indonesia', 'Pakistan', 'Mexico', 'Russia', 'Turkey',
       'Australia', 'Nigeria', 'Greece', 'Belgium', 'Japan', 'Egypt',
       'Singapore', 'Brazil', 'Poland', 'China',
       'Iran, Islamic Republic of...', 'United States of America',
       'Italy', 'Viet Nam', 'Israel', 'Peru', 'South Africa', 'Other',
       'Spain', 'Bangladesh',
       'United Kingdom of Great Britain and Northern Ireland', 'France',
       'Switzerland', 'Algeria', 'Tunisia', 'Argentina', 'Sweden',
       'Colombia', 'I do not wish to disclose my location', 'Canada',
       'Chile', 'Netherlands', 'Ukraine', 'Saudi Arabia', 'Romania',
       'Morocco', 'Austria', 'Taiwan', 'Kenya', 'Belarus', 'Ireland',
       'Portugal', 'Hong Kong (S.A.R.)', 'Denmark', 'Germany',
       'South Korea', 'Philippines', 'Sri Lanka', 'United Arab Emirates',
       'Uganda', 'Ghana', 'Malaysia', 'Thailand', 'Nepal', 'Kazakhstan',
       'Ethiopia', 'Iraq', 'Ecuador', 'Norway', 'Czech Republic'],
      dtype=obje

Para la columna 'country' observamos que:

- queremos cambiar el valor 'Iran, Islamic Republic of...' por 'Islamic Republic of Iran'

- el valor 'Viet Nam' está mal escrito, ya que se escribe junto


In [853]:
df['current_role'].unique()

array(['Other', 'Program/Project Manager', 'Software Engineer',
       'Research Scientist', 'Currently not employed', 'Student',
       'Data Scientist', 'Data Analyst', 'Machine Learning Engineer',
       'Business Analyst', 'Data Engineer', 'Product Manager',
       'Statistician', 'Developer Relations/Advocacy',
       'DBA/Database Engineer'], dtype=object)

Basaremos nuestro análisis en el perfil de 'Data Analyst'

In [854]:
df['yearly_compensation'].unique()

array(['25,000-29,999', '60,000-69,999', '$0-999', '30,000-39,999', nan,
       '15,000-19,999', '70,000-79,999', '2,000-2,999', '10,000-14,999',
       '5,000-7,499', '20,000-24,999', '1,000-1,999', '100,000-124,999',
       '7,500-9,999', '4,000-4,999', '40,000-49,999', '50,000-59,999',
       '3,000-3,999', '300,000-499,999', '200,000-249,999',
       '125,000-149,999', '250,000-299,999', '80,000-89,999',
       '90,000-99,999', '150,000-199,999', '>$1,000,000',
       '$500,000-999,999'], dtype=object)

Encontramos 1 símbolo extraño que habría que eliminar: `$`. Además existen valores nulos

In [855]:
df['programming_languages'].unique()

array(['Python, R', ' SQL, C, C++, Java', 'Python, C++, Java', 'Python',
       'Python, C, MATLAB', ' C++, Java, Javascript', 'Python, SQL',
       'Python, R, SQL', 'Python, R, C++', 'Python, MATLAB',
       'Python, C, C++, MATLAB', 'Python, SQL, C, Java, Javascript',
       ' None', 'Python, R, SQL, C++', ' R', 'Python, SQL, C++',
       'Python, R, SQL, C, C++, MATLAB', 'Python, C++', 'Python, C, C++',
       'Python, Java', 'Python, C, C++, Java, Javascript, MATLAB',
       'Python, Java, Javascript', 'Python, SQL, Java, Other',
       'Python, SQL, Java', 'Python, C, C++, Other',
       'Python, C++, MATLAB',
       'Python, SQL, C, C++, Java, Javascript, MATLAB',
       'Python, SQL, Java, Javascript, MATLAB, Other', ' R, SQL',
       'Python, SQL, Other', ' SQL, Javascript, Bash', ' C++, Other',
       'Python, SQL, C', ' C++', 'Python, SQL, Bash',
       'Python, C, Javascript', 'Python, Javascript, Bash, MATLAB',
       'Python, Javascript', ' SQL, Javascript', ' SQL, Java',

In [856]:
df['programming_languages'].iloc[:2]

0             Python, R
1     SQL, C, C++, Java
Name: programming_languages, dtype: object

A simple vista parece que no hay errores en la escritura, aunque lo comprobaremos más adelante cuando apliquemos los métodos split y explode. 

In [857]:
df['ides'].unique()

array([' Vim / Emacs', ' Notepad++,Jupyter Notebook',
       'PyCharm ,Jupyter Notebook, Other', ...,
       'Visual Studio ,Visual Studio Code (VSCode) , Notepad++, Vim / Emacs,MATLAB ,Jupyter Notebook',
       'RStudio ,Visual Studio ,PyCharm , Spyder',
       'RStudio ,Visual Studio ,Visual Studio Code (VSCode) , Notepad++, Vim / Emacs'],
      dtype=object)

A simple vista observamos que no existen errores en la escritura

In [858]:
df['visualization'].unique()

array([' Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ,Leaflet / Folium ',
       ' Matplotlib ', ' Matplotlib ,Seaborn ,Ggplot / ggplot2 ',
       'Geoplotlib ', ' Matplotlib ,Plotly / Plotly Express ',
       ' Matplotlib ,Seaborn ', nan, 'Plotly / Plotly Express ',
       ' Matplotlib ,Seaborn ,Plotly / Plotly Express ', ' None',
       'Ggplot / ggplot2 ', ' Matplotlib ,Seaborn , Other',
       ' Matplotlib ,Seaborn ,Plotly / Plotly Express ,Ggplot / ggplot2 ,Shiny ',
       'Altair ', ' Matplotlib ,Ggplot / ggplot2 ',
       ' Matplotlib ,Seaborn ,Bokeh ',
       ' Matplotlib ,Seaborn ,Plotly / Plotly Express ,Ggplot / ggplot2 ',
       'Seaborn ,Ggplot / ggplot2 ,Shiny ',
       'Plotly / Plotly Express ,Ggplot / ggplot2 ,Shiny ',
       ' Matplotlib ,Plotly / Plotly Express ,Ggplot / ggplot2 ,D3 js ',
       ' Matplotlib ,Seaborn ,Plotly / Plotly Express ,Bokeh ',
       ' Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Geoplotlib ', ' Other',
       'Seaborn ', ' Matplotlib ,Altair ', '

A simple vista parece que no hay errores en la escritura, aunque lo comprobaremos más adelante cuando apliquemos los métodos split y explode. 

In [859]:
df['activities'].unique()

array([' None of these activities are an important part of my role at work',
       ' Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data',
       ' Do research that advances the state of the art of machine learning',
       'Analyze and understand data to influence product or business decisions, Build prototypes to explore applying machine learning to new areas',
       nan,
       ' Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data, Build prototypes to explore applying machine learning to new areas',
       'Analyze and understand data to influence product or business decisions',
       'Analyze and understand data to influence product or business decisions, Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data',
       ' Build and/or run the data infrastructure that my business uses for storing, analyzi

Observamos que habrá que eliminar espacios al inicio de algunos strings y que además dependiendo del usuario, se ha elegido una o varias opciones. Por tanto habrá que usar un split y explode

In [860]:
df['age'].unique()

array(['50-54', '22-24', '45-49', '25-29', '18-21', '30-34', '40-44',
       '35-39', '70+', '55-59', '60-69'], dtype=object)

Nos llama la atención que existan usuarios con 70 años o más 

In [861]:
df['gender'].unique()

array(['Man', 'Woman', 'Nonbinary', 'Prefer not to say',
       'Prefer to self-describe'], dtype=object)

In [862]:
df['primary_tool'].unique()

array(['Local development environments (RStudio, JupyterLab, etc.)',
       'Advanced statistical software (SPSS, SAS, etc.)',
       'Basic statistical software (Microsoft Excel, Google Sheets, etc.)',
       nan,
       'Business intelligence software (Salesforce, Tableau, Spotfire, etc.)',
       'Cloud-based data software & APIs (AWS, GCP, Azure, etc.)',
       'Other'], dtype=object)

Hemos detectado que hay valores nulos, habrá que indagar en ellos.

In [863]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25973 entries, 0 to 25972
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   country                25973 non-null  object
 1   current_role           25973 non-null  object
 2   yearly_compensation    15391 non-null  object
 3   programming_languages  24941 non-null  object
 4   ides                   24934 non-null  object
 5   visualization          23970 non-null  object
 6   activities             15635 non-null  object
 7   age                    25973 non-null  object
 8   gender                 25973 non-null  object
 9   primary_tool           18489 non-null  object
dtypes: object(10)
memory usage: 2.0+ MB


Nuestro dataframe tiene 25973 filas y 10 columnas. Todas nuestras columnas son de tipo object y yambién observamos que tenemos algunas columnas con bastantes valores nulos 

In [864]:
df.describe().T

count unique  \
country                25973     66   
current_role           25973     15   
yearly_compensation    15391     26   
programming_languages  24941    825   
ides                   24934   1330   
visualization          23970    467   
activities             15635     96   
age                    25973     11   
gender                 25973      5   
primary_tool           18489      6   

                                                                     top  \
country                                                            India   
current_role                                                     Student   
yearly_compensation                                               $0-999   
programming_languages                                             Python   
ides                                                    Jupyter Notebook   
visualization                                       Matplotlib ,Seaborn    
activities             Analyze and understand data to influence produ...   
age                                                                25-29   
gender                                                               Man   
primary_tool           Basic statistical software (Microsoft Excel, G...   

                        freq  
country                 7434  
current_role            6804  
yearly_compensation     3369  
programming_languages   4740  
ides                    2341  
visualization           5697  
activities              2891  
age                     4931  
gender                 20598  
primary_tool            7246

Podemos extraer mucha información de este método, como por ejemplo que los usuarios de India son los que más han participado en esta encuesta, o que el sueldo que más se repite es el que va de 0 a 999. Por otro lado, el lenguaje de programación más utilizado es Python. También nos llama la atención que la mayoría de puestos sea el de estudiantes

In [865]:
pd.DataFrame(df.isnull().sum() * 100 / df.shape[0])

0
country                 0.000000
current_role            0.000000
yearly_compensation    40.742309
programming_languages   3.973357
ides                    4.000308
visualization           7.711855
activities             39.802872
age                     0.000000
gender                  0.000000
primary_tool           28.814538

Observamos que en las columnas que más nulos tienen son 'yearly_compensation' con un 40,74% y 'activities' con un 39.80%. No queremos eliminarlos ya que son una parte importante de nuestro análisis.

# LIMPIEZA: 


Para la columna 'country':

- cambiar el valor 'Iran, Islamic Republic of...' por 'Islamic Republic of Iran'

- el valor 'Viet Nam' está mal escrito, ya que se escribe junto

In [866]:
def limpiar_country(columna):
    """
    Esta función nos permite modificar los datos de una columna para que estén escritos adecuadamente
    Args: columna
    columna(string): es el dato que queremos modificar 
    """
    
    if columna == 'Iran, Islamic Republic of...':
        cambiado = columna.replace('Iran, Islamic Republic of...', 'Islamic Republic of Iran')
        return cambiado
    
    elif columna == 'Viet Nam':
        cambiado2 = columna.replace('Viet Nam', 'Vietnam')
        return cambiado2
    
    else: 
        return columna

In [867]:
df['country'] = df['country'].apply(limpiar_country)
df['country'].unique()

array(['India', 'Indonesia', 'Pakistan', 'Mexico', 'Russia', 'Turkey',
       'Australia', 'Nigeria', 'Greece', 'Belgium', 'Japan', 'Egypt',
       'Singapore', 'Brazil', 'Poland', 'China',
       'Islamic Republic of Iran', 'United States of America', 'Italy',
       'Vietnam', 'Israel', 'Peru', 'South Africa', 'Other', 'Spain',
       'Bangladesh',
       'United Kingdom of Great Britain and Northern Ireland', 'France',
       'Switzerland', 'Algeria', 'Tunisia', 'Argentina', 'Sweden',
       'Colombia', 'I do not wish to disclose my location', 'Canada',
       'Chile', 'Netherlands', 'Ukraine', 'Saudi Arabia', 'Romania',
       'Morocco', 'Austria', 'Taiwan', 'Kenya', 'Belarus', 'Ireland',
       'Portugal', 'Hong Kong (S.A.R.)', 'Denmark', 'Germany',
       'South Korea', 'Philippines', 'Sri Lanka', 'United Arab Emirates',
       'Uganda', 'Ghana', 'Malaysia', 'Thailand', 'Nepal', 'Kazakhstan',
       'Ethiopia', 'Iraq', 'Ecuador', 'Norway', 'Czech Republic'],
      dtype=object)

Hemos aplicado un unique para comprobar que los cambios se han realizado correctamente

Para la columna 'yearly_compensation': 
-  símbolos extraños a eliminar: `$`    

In [868]:
def limpiar_yearly(columna):
    
    """
    Esta función nos permite eliminar el símbolo $ de los datos de nuestras columnas
    Args: columna
    columna(string): es el dato que queremos modificar 
    """
    try:
        return re.sub('\$', '', columna)
         
    except:
        return columna
    

In [869]:
df['yearly_compensation'] = df['yearly_compensation'].apply(limpiar_yearly)
df['yearly_compensation'].unique()

array(['25,000-29,999', '60,000-69,999', '0-999', '30,000-39,999', nan,
       '15,000-19,999', '70,000-79,999', '2,000-2,999', '10,000-14,999',
       '5,000-7,499', '20,000-24,999', '1,000-1,999', '100,000-124,999',
       '7,500-9,999', '4,000-4,999', '40,000-49,999', '50,000-59,999',
       '3,000-3,999', '300,000-499,999', '200,000-249,999',
       '125,000-149,999', '250,000-299,999', '80,000-89,999',
       '90,000-99,999', '150,000-199,999', '>1,000,000',
       '500,000-999,999'], dtype=object)

Hemos aplicado un unique para comprobar que los cambios se han realizado correctamente

Para la columna 'activities':
- eliminar espacios al inicio de algunos strings

In [870]:
def limpiar_activities(columna):
    """
    Esta función nos permite eliminar los espacios que hay al principio y al final de los strings 
    Args: columna
    columna(string): es el dato al que le queremos quitar los espacios 
    """
    try:
        return columna.strip()
         
    except:
        return columna

In [871]:
df['activities'] = df['activities'].apply(limpiar_activities)
df['activities'].unique()

array(['None of these activities are an important part of my role at work',
       'Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data',
       'Do research that advances the state of the art of machine learning',
       'Analyze and understand data to influence product or business decisions, Build prototypes to explore applying machine learning to new areas',
       nan,
       'Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data, Build prototypes to explore applying machine learning to new areas',
       'Analyze and understand data to influence product or business decisions',
       'Analyze and understand data to influence product or business decisions, Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data',
       'Build and/or run the data infrastructure that my business uses for storing, analyzing, a

Hemos aplicado un unique para comprobar que los cambios se han realizado correctamente

A partir de ahora, basaremos nuestro análisis en aquellos usuarios cuyo trabajo es 'Data Analyst'

In [872]:
############################################### Business Analyst

df_analyst = df[df['current_role'] == 'Data Analyst']

In [873]:
df_analyst['current_role'].unique()

array(['Data Analyst'], dtype=object)

Continuamos limpiando nuestro dataframe, buscaremos los valores duplicados y los nulos

In [874]:
#Buscamos los duplicados
df_analyst.duplicated().sum()

4

In [875]:
df_analyst[df_analyst.duplicated()].head(2)

country  current_role yearly_compensation programming_languages ides  \
8178   India  Data Analyst               0-999                   NaN  NaN   
9196   India  Data Analyst         7,500-9,999                   NaN  NaN   

     visualization                                         activities    age  \
8178           NaN  Analyze and understand data to influence produ...  25-29   
9196           NaN  Analyze and understand data to influence produ...  25-29   

     gender                                       primary_tool  
8178    Man  Basic statistical software (Microsoft Excel, G...  
9196    Man  Basic statistical software (Microsoft Excel, G...

In [876]:
#Eliminamos los valores duplicados
df_analyst.drop_duplicates(inplace = True)

In [877]:
df_analyst.duplicated().sum()

0

In [878]:
#Buscamos los valores nulos
pd.DataFrame((df_analyst.isnull().sum() * 100) / df_analyst.shape[0])

0
country                 0.000000
current_role            0.000000
yearly_compensation    11.406182
programming_languages   6.094906
ides                    6.225512
visualization           8.271659
activities              9.882455
age                     0.000000
gender                  0.000000
primary_tool           23.160644

In [879]:
#FUNCIOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOON

df_analyst['yearly_compensation'].replace(np.nan, 'unknow', inplace = True)

In [880]:
df_analyst['programming_languages'].replace(np.nan, 'unknow', inplace = True)

In [881]:
df_analyst['ides'].replace(np.nan, 'unknow', inplace = True)

In [882]:
df_analyst['visualization'].replace(np.nan, 'unknow', inplace = True)

In [883]:
df_analyst['activities'].replace(np.nan, 'unknow', inplace = True)

In [884]:
df_analyst['primary_tool'].replace(np.nan, 'unknow', inplace = True)

In [885]:
pd.DataFrame((df_analyst.isnull().sum() * 100) / df_analyst.shape[0])

0
country                0.0
current_role           0.0
yearly_compensation    0.0
programming_languages  0.0
ides                   0.0
visualization          0.0
activities             0.0
age                    0.0
gender                 0.0
primary_tool           0.0

Como observamos, hemos reemplazado todos los valores nulos de nuestro dataframe por la palabra 'unknown' y de esta forma no perdemos los datos y podemos seguir trabajando con ellos. 

In [886]:
df_analyst.head(2)

country  current_role yearly_compensation  \
16                   Belgium  Data Analyst         2,000-2,999   
32  United States of America  Data Analyst              unknow   

   programming_languages                                               ides  \
16           Python, SQL  Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
32                     R                                           RStudio    

                                     visualization  \
16   Matplotlib ,Seaborn ,Plotly / Plotly Express    
32                               Ggplot / ggplot2    

                                           activities    age     gender  \
16  Analyze and understand data to influence produ...  50-54        Man   
32                                             unknow  22-24  Nonbinary   

                                         primary_tool  
16  Local development environments (RStudio, Jupyt...  
32                                             unknow

### COMENZAMOS NUESTRO ANALISIS

Empezamos a categorizar nuestras columnas. Para ello creamos una función 

In [887]:
def rango_edad(col):
    
    joven = ['25-29', '18-21', '22-24']
    mediana =['30-34', '35-39', '40-44', '45-49', '50-54']
    mayor = ['55-59','60-69', '70+']
    
    if col in joven:
        return 'joven'
        
    elif col in mediana:
        return 'mediana'
        
    elif col in mayor:
        return 'mayor'

In [888]:
df_analyst['ages'] = df_analyst['age'].apply(rango_edad)

In [889]:
df_analyst['ages'].unique()

array(['mediana', 'joven', 'mayor'], dtype=object)

In [890]:
df_analyst.head()

country  current_role yearly_compensation  \
16                   Belgium  Data Analyst         2,000-2,999   
32  United States of America  Data Analyst              unknow   
33                     Egypt  Data Analyst         7,500-9,999   
46                     China  Data Analyst               0-999   
52              South Africa  Data Analyst       25,000-29,999   

                            programming_languages  \
16                                    Python, SQL   
32                                              R   
33                                         Python   
46  Python, SQL, C, C++, Java, Javascript, MATLAB   
52                                         R, SQL   

                                                 ides  \
16  Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
32                                           RStudio    
33                         Notepad++,Jupyter Notebook   
46  Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
52                          RStudio ,Jupyter Notebook   

                                        visualization  \
16      Matplotlib ,Seaborn ,Plotly / Plotly Express    
32                                  Ggplot / ggplot2    
33                                               None   
46               Matplotlib ,Plotly / Plotly Express    
52  Plotly / Plotly Express ,Ggplot / ggplot2 ,Shiny    

                                           activities    age     gender  \
16  Analyze and understand data to influence produ...  50-54        Man   
32                                             unknow  22-24  Nonbinary   
33  Analyze and understand data to influence produ...  30-34      Woman   
46  Analyze and understand data to influence produ...  22-24        Man   
52  Analyze and understand data to influence produ...  40-44        Man   

                                         primary_tool     ages  
16  Local development environments (RStudio, Jupyt...  mediana  
32                                             unknow    joven  
33  Basic statistical software (Microsoft Excel, G...  mediana  
46  Basic statistical software (Microsoft Excel, G...    joven  
52  Local development environments (RStudio, Jupyt...  mediana

Comprobamos que los cambios se han realizado correctamente, por tanto eliminamos la columna original. 

In [891]:
df_analyst.drop('age', inplace = True, axis = 1)

## Hacemos split/explode para la columna 'activities'

In [ ]:
df_analyst.head()

In [ ]:
df_analyst.iloc[3,6]

In [ ]:
df_analyst['activities'].unique()

Para nuestro análisis, necesitamos acceder a las actividades que un Data Analyst realuza en su día a día

In [ ]:
df_analyst['activities'].unique()

In [ ]:
patron = ',\s[A-Z]'

In [ ]:
df_analyst['activities'] = df_analyst['activities'].str.split(patron, regex = True, n = -1)

In [ ]:
df_prueba = df_analyst.explode('activities')

In [892]:
def crear_filas(df,patron,columna):
    
    df[columna] = df[columna].str.split(patron, n = -1) 
    
    return df.explode(columna)

In [893]:
df_final = crear_filas(df_analyst, ', ', 'activities')
df_final 

country  current_role yearly_compensation  \
16                      Belgium  Data Analyst         2,000-2,999   
32     United States of America  Data Analyst              unknow   
33                        Egypt  Data Analyst         7,500-9,999   
46                        China  Data Analyst               0-999   
46                        China  Data Analyst               0-999   
...                         ...           ...                 ...   
25966                     India  Data Analyst         3,000-3,999   
25968                     Egypt  Data Analyst       15,000-19,999   
25968                     Egypt  Data Analyst       15,000-19,999   
25968                     Egypt  Data Analyst       15,000-19,999   
25968                     Egypt  Data Analyst       15,000-19,999   

                               programming_languages  \
16                                       Python, SQL   
32                                                 R   
33                                            Python   
46     Python, SQL, C, C++, Java, Javascript, MATLAB   
46     Python, SQL, C, C++, Java, Javascript, MATLAB   
...                                              ...   
25966                                      Python, C   
25968                  Python, SQL, Javascript, Bash   
25968                  Python, SQL, Javascript, Bash   
25968                  Python, SQL, Javascript, Bash   
25968                  Python, SQL, Javascript, Bash   

                                                    ides  \
16     Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
32                                              RStudio    
33                            Notepad++,Jupyter Notebook   
46     Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
46     Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
...                                                  ...   
25966                          PyCharm ,Jupyter Notebook   
25968  Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
25968  Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
25968  Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   
25968  Jupyter (JupyterLab, Jupyter Notebooks, etc) ,...   

                                        visualization  \
16      Matplotlib ,Seaborn ,Plotly / Plotly Express    
32                                  Ggplot / ggplot2    
33                                               None   
46               Matplotlib ,Plotly / Plotly Express    
46               Matplotlib ,Plotly / Plotly Express    
...                                               ...   
25966          Matplotlib ,Seaborn ,Ggplot / ggplot2    
25968                            Matplotlib ,Seaborn    
25968                            Matplotlib ,Seaborn    
25968                            Matplotlib ,Seaborn    
25968                            Matplotlib ,Seaborn    

                                              activities     gender  \
16     Analyze and understand data to influence produ...        Man   
32                                                unknow  Nonbinary   
33     Analyze and understand data to influence produ...      Woman   
46     Analyze and understand data to influence produ...        Man   
46     Build and/or run the data infrastructure that ...        Man   
...                                                  ...        ...   
25966  Experimentation and iteration to improve exist...        Man   
25968  Analyze and understand data to influence produ...        Man   
25968  Build and/or run the data infrastructure that ...        Man   
25968                                          analyzing        Man   
25968                          and operationalizing data        Man   

                                            primary_tool     ages  
16     Local development environments (RStudio, Jupyt...  mediana  
32                                                unknow    joven  
33     Basic statistical software (Microsoft Excel, G...  m

In [894]:
df_final['activities'].unique()

array(['Analyze and understand data to influence product or business decisions',
       'unknow',
       'Build and/or run the data infrastructure that my business uses for storing',
       'analyzing', 'and operationalizing data',
       'Build prototypes to explore applying machine learning to new areas',
       'Build and/or run a machine learning service that operationally improves my product or workflows',
       'None of these activities are an important part of my role at work',
       'Experimentation and iteration to improve existing ML models',
       'Do research that advances the state of the art of machine learning',
       'Other'], dtype=object)

In [896]:
def limpiar_actividades(columna): 
    if columna == 'Build and/or run the data infrastructure that my business uses for storing':
        return 'Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data'
    
    elif columna == 'analyzing':
        return 'Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data'
    
    elif columna == 'and operationalizing data':
        return 'Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data'
    
    else:
        return columna

In [898]:
df_final['activities'] = df_final['activities'].apply(limpiar_actividades)

In [899]:
df_final['activities'].unique()

array(['Analyze and understand data to influence product or business decisions',
       'unknow',
       'Build and/or run the data infrastructure that my business uses for storing, analyzing, and operationalizing data',
       'Build prototypes to explore applying machine learning to new areas',
       'Build and/or run a machine learning service that operationally improves my product or workflows',
       'None of these activities are an important part of my role at work',
       'Experimentation and iteration to improve existing ML models',
       'Do research that advances the state of the art of machine learning',
       'Other'], dtype=object)

In [901]:
df_final['ides'].unique()

array(['Jupyter (JupyterLab, Jupyter Notebooks, etc) ,Jupyter Notebook',
       'RStudio ', ' Notepad++,Jupyter Notebook',
       'Jupyter (JupyterLab, Jupyter Notebooks, etc) ,Visual Studio ,Visual Studio Code (VSCode) ,PyCharm ,MATLAB ,Jupyter Notebook',
       'RStudio ,Jupyter Notebook', 'PyCharm , Spyder', ' None',
       'Jupyter Notebook, Other', 'Visual Studio ,Jupyter Notebook',
       'Jupyter Notebook',
       'Jupyter (JupyterLab, Jupyter Notebooks, etc) ,Visual Studio Code (VSCode) ,Jupyter Notebook',
       'Jupyter (JupyterLab, Jupyter Notebooks, etc) ,Visual Studio ,PyCharm , Notepad++,Jupyter Notebook',
       'Jupyter (JupyterLab, Jupyter Notebooks, etc) ,RStudio , Spyder',
       ' Spyder',
       'Jupyter (JupyterLab, Jupyter Notebooks, etc) ,Visual Studio Code (VSCode) ,PyCharm ,Jupyter Notebook',
       'Visual Studio ', 'PyCharm ,Jupyter Notebook',
       'Jupyter (JupyterLab, Jupyter Notebooks, etc) ,RStudio ,PyCharm , Spyder, Sublime Text,Jupyter Notebook',
   

## Hacemos split/explode para la columna 'activities'

# Preguntas: 


1. ¿Cuáles son los países en los que más se gana y en los que menos? ¿En qué puesto se encuentra España? 

2. ¿Existe relación entre el rango de edad y el género? (en general, no solo para data analyst)

   2.1 nuestra hipotesis es que a medida que pasan los años, la brecha de género se acentúa 

3. ¿Existe relación entre el salario y la ocupación? El sueldo que más se repite es de 0-999 y pensamos que tal vez sea porque contratan estudiantes como becarios



In [ ]:
tabla_completa.head(1)

In [ ]:
tabla_completa['GENDER'].unique()

In [ ]:
plt.figure(figsize = (10,5))
sns.countplot(x = 'GENDER', data = tabla_completa, hue = 'GENDER');

In [ ]:
df_woman = tabla_completa[tabla_completa["GENDER"] == "Woman"]
df_woman.head()

In [ ]:
df_woman["AGE"].unique()

In [ ]:
df_edad = df_woman["AGE"].value_counts().reset_index().sort_values(by = ["AGE"])
df_edad

In [ ]:
def rango(col):
    joven = ['25-29', '18-21', '22-24']
    mediana =['30-34', '35-39', '40-44', '45-49', '50-54']
    mayor = ['55-59','60-69', '70+']

    if col in joven:
        return 'joven'
    elif col in mediana:
        return 'mediana'
    else:
        return 'mayor'
    
    
     

In [ ]:
rango("18-21")

In [ ]:
tabla_completa["EDAD_PRUEBA"] = tabla_completa["AGE"].apply(rango)
tabla_completa.head()

In [ ]:
plt.bar(x = df_edad["AGE"], height = df_edad["count"]);

In [ ]:
df_man = tabla_completa[tabla_completa["GENDER"] == "Man"]
df_man.head()

In [ ]:
df_man["AGE"].unique()

In [ ]:
df_edad2 = df_man["AGE"].value_counts().reset_index().sort_values(by = ["AGE"])
df_edad2

In [ ]:
plt.bar(x = df_edad2["AGE"], height = df_edad["count"]);

In [ ]:
sns.swarmplot(x = "AGE", y = "count", data = df_edad);